clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -4148,18 +4226,21 @@ void Sema::ConstructOpenCLKernel(FunctionDecl *KernelCallerFunc,
     // Attributes of a user-written SYCL kernel must be copied to the internally
     // generated alternative kernel, identified by a known string in its name.
     if (StableName.find("__pf_kernel_wrapper") != std::string::npos)
-      copySYCLKernelAttrs(KernelObj);
+      copySYCLKernelAttrs(KernelObj, KernelCallerFunc);
   }
 
-  bool IsSIMDKernel = isESIMDKernelType(KernelObj);
+  KernelCallOperatorVisitor KernelCallOperator(KernelCallerFunc, KernelObj);
+
+  bool IsSIMDKernel =
+      isESIMDKernelType(KernelCallOperator, KernelObj, KernelCallerFunc, *this);
 
   SyclKernelDeclCreator kernel_decl(*this, KernelObj->getLocation(),
                                     KernelCallerFunc->isInlined(), IsSIMDKernel,
                                     KernelCallerFunc);
   SyclKernelBodyCreator kernel_body(*this, kernel_decl, KernelObj,
                                     KernelCallerFunc);
   SyclKernelIntHeaderCreator int_header(
-      *this, getSyclIntegrationHeader(), KernelObj,
+      KernelCallOperator, *this, getSyclIntegrationHeader(), KernelObj,
```

Comment:
We can just pass IsSIMDKernel here instead since we've already calculated it. I'm also wondering if it makes sense to get the 'correct' call and pass it to whatever visitor needs it/store this information in some base class, instead of AST traversal multiple times. WDYT @Fznamznon  @premanandrao